### PR TITLE
Basic: various improvements from Geany

### PR DESCRIPTION
**Pull request by techee**
There were some improvements in the Geany parser, mostly related to extraction of variables in `dim` statements.

Also, the uctags parser determined the used basic dialect based on file extension which wouldn't work very well for Geany as it parses files using MIO buffer. What the parser did appears quite a non-standard approach with the file extensions and I haven't seen it in any other parser. What I did is that I basically just merged all the basic dialect keywords into one and allowed both labels like `label:` and `.label` for any kind of file - hope it's OK.

I added the `dim` extraction code as it was in Geany and then made a few commits on top of that as there were some minor problems in the code and finally made a commit to cleanup the code a bit.

### CODE REVIEW

File: parsers/basic.c [hunk 1]
```diff
@@ -83,6 +82,83 @@ static KeyWord basic_keywords[] = {
...
+{
+	const char *old_pos = pos;
+	while (isspace (*pos))
+		pos++;
+
+	/* create tags only if there is some space between the keyword and the identifier */
+	if (old_pos == pos)
+		return 0;
+
+	vStringClear (name);
```
Comment (masatake) [inline parsers/basic.c]: How about moving this vStringClear to ...

File: parsers/basic.c [hunk 2]
```diff
@@ -83,6 +82,83 @@ static KeyWord basic_keywords[] = {
...
+			pos++;
+	}
+	/*	capture "dim as string * 4096 chunk" */
+	if (strncmp (pos, "*", 1) == 0)
+	{
+		pos += 1; /* skip "*" */
+		while (isspace (*pos) || isdigit(*pos) || ispunct(*pos))
+			pos++;
+	}
+
```
Comment (masatake) [inline parsers/basic.c:122]: ...here?

File: parsers/basic.c [hunk 3]
```diff
@@ -83,6 +82,83 @@ static KeyWord basic_keywords[] = {
...
+		if (*pos == '\'')
+			return 0; /* break if we are in a comment */
+
+		vStringClear (name);
+		for (; *pos && !isspace (*pos) && *pos != '(' && *pos != ',' && *pos != '='; pos++)
+			vStringPut (name, *pos);
+		makeSimpleTag (name, kind);
+	}
+
+	vStringDelete (name);
```
Comment (masatake) [inline parsers/basic.c:149]: Is it o.k. to delete `name` here?
I think this deletion should be at ...(B)

File: parsers/basic.c [hunk 4]
```diff
@@ -109,6 +185,9 @@ static int match_keyword (const char *p, KeyWord const *kw)
 	name = vStringNew ();
 	p += i;
 
+	if (kw->kind == K_VARIABLE)
+		return extract_dim (p, name, kw->kind); /* extract_dim adds the found tag(s) */
```
Comment (masatake) [inline parsers/basic.c]: (B)... here like:
```
if (kw->kind == K_VARIABLE)
{
    int r = extract_dim (p, name, kw->kind);
    vStringDelete (name);
    return r;
}

File: parsers/basic.c [hunk 5]
```diff
@@ -140,13 +140,13 @@ static int extract_dim (char const *pos, vString * name, BasicKind kind)
 			pos++;
 
 		if (*pos == '\'')
-			return 0; /* break if we are in a comment */
+			break; /* break if we are in a comment */
```
Comment (masatake) [inline parsers/basic.c:135]: You made a separate commit for these changes.
However, I think this change should be part of the commit " basic: add more advanced extraction of dim-like variables".

If you extend the commit introducing `extract_dim`,  or add a new feature to the commit, making a new commit is greatly appreciated. However, it seems that these changes fix bugs in the base commit, " basic: add more advanced extraction of dim-like variables". In such a case, the commits should be unified.

File: parsers/basic.c [hunk 6]
```diff
@@ -114,7 +114,7 @@ static int extract_dim (char const *pos, vString * name, BasicKind kind)
 		/* now we are at the name */
 	}
 	/* capture "dim as foo ptr bar" */
-	if (strncasecmp (pos, "ptr", 3) == 0 && isspace(*(pos+4)))
+	if (strncasecmp (pos, "ptr", 3) == 0 && isspace(*(pos+3)))
```
Comment (masatake) [inline parsers/basic.c:109]: I think this change should be merged to the commit " basic: add more advanced extraction of dim-like variables".

File: parsers/basic.c (hunk 3 above)
Comment (masatake) [reply parsers/basic.c:149]: After reading the last commit, putting `vStringDelete (name);` here makes sense.
However, I don't like these commit sequences. 

In this commit, I think you did two things: (1) fixing the memory management bug in extract_dim introduced in "basic: add more advanced extraction of dim-like variables", and (2) reorganizing the way to manage a vString storing "name".

(1) should be merged to  "basic: add more advanced extraction of dim-like variables".
So you can focus on (2) in this commit.

I think what I wrote here was a  kind of "Preaching to Buddha".

**Comment by masatake** [comment]
The result of the changes looks good.
However, I think the sequence of commits can be improved.

**Comment by masatake** [comment]
I must say "thank you" first of all.

Do you have any real code written in Basic?
As you may know, I'm gathering "real code" as corpus at https://github.com/universal-ctags/codebase.
I evaluated the improvement for Markdown with the coebase.

I'm using `size:>100000  language:BASIC` as a search expression at Github's code search.

What I found was:
https://github.com/jonatasdemello/gwbasic

If you have any recommendations, I would like to know them.

**Comment by techee** [comment]
> However, I think the sequence of commits can be improved.

It's because I first tried to add the code we had in Geany - which was up to the commit `basic: remove now unused 'skip' KeyWord member` and afterwards started fixing the problems I noticed (which were also present in the Geany code). I'll try to improve it a bit.

> Do you have any real code written in Basic?

Nope, this is my first experience with Basic and hope the last ;-)

Apart from `c.c`, this is probably all we had in Geany that could be interesting for ctags. There's also this Matlab parser

https://github.com/geany/geany/blob/master/ctags/parsers/geany_matlab.c

which doesn't extract all the kinds like the regex parser that ctags uses. Right now I'm not very motivated to improve this parser myself (and we'll probably keep this parser for now because of performance reasons).

**Comment by techee** [comment]
@masatake I hope it's better now.

**Comment by techee** [comment]
I added a fix of this issue we had in Geany: https://github.com/geany/geany/issues/2461

**Review by masatake — APPROVED** [review]

**Comment by masatake** [comment]
Thank you very much. Looks fine. I will merge soon.

**Pull request merged**